conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
+                                # every gpu, even if all the data is on the
+                                # same gpu.  The workaround:  construct the 
+                                # sparse tensor on the cpu then move it.
+                                ctor_cpu = eval(
+                                    re.sub('\.cuda', '', d_p.type()))
+                                sparse_data = ctor_cpu(
+                                    d_p._indices().cpu(), vals.cpu(),
+                                    d_p.size()).to(d_p.device)
+                            d_p.add_(weight_decay, sparse_data)
+                            del flat_indices, sparse_data, vals
+                if momentum != 0:
+                    param_state = self.state[p]
+                    if 'momentum_buffer' not in param_state:
+                        buf = param_state['momentum_buffer'] = torch.zeros_like(d_p)
+                        buf.mul_(momentum).add_(d_p)
+                    else:
+                        buf = param_state['momentum_buffer']
+                        buf.mul_(momentum).add_(1 - dampening, d_p)
+                    if d_p_is_sparse:
+                        buf = buf.coalesce()
+                        param_state['momentum_buffer'] = buf
+                    if nesterov:
+                        d_p = d_p.add(momentum, buf)
+                    else:
+                        d_p = buf
+                if d_p_is_sparse:
+                    d_p = d_p.coalesce()
+                    p.grad.data = d_p
+
+                p.data.add_(-group['lr'], d_p)
+
+        return loss
+
+
+class DataParallelizedModule(nn.Module):
+    """
+    Similarly to nn.DataParallel, this class of modules serves to wrap 
+    other modules and run them in parallel over multiple gpus, splitting 
+    training (or testing/application) batches between the gpus (over the 
+    first dimension of the batch).  
+    
+    Unlike nn.DataParallel, all of the wrapped module's parameters will be 
+    copied to all gpus during construction of the wrapper, and only gradient 
+    data will be copied from gpu to gpu during training (no data should need 
+    to be copied between gpus during forward application).  However, training 
+    should only use loss functions that can be expressed as averages over all 
+    data points of a batch of some per-data-point loss.  Also, the set of 
+    parameters of the wrapped module is assumed not to change (of course their 
+    values can change) over the lifetime of the wrapper object.
+    
+    Note that during training of a wrapped module, it is necessary to call 
+    the wrapper's broadcast_gradients method immediately following the 
+    backpropagation of gradients (i.e. typically right after calling 
+    .backward on some loss tensor), in order to share corrections to the 
+    computed gradients between the gpus.
+    """
+    def __init__(self, module, device, copy_fn=copy.deepcopy):
```

Comment:
Why should `copy_fn` be a parameter defaulting to deepcopy? Deepcopy makes me nervous in general, because I'd rather know what we intend to copy; also, it appears that deepcopy isn't the right thing in the one place this is used.

Reply:
In general I'd say the right way to copy a torch.nn.Module would be via its state_dict.  It's annoying that that won't quite work in the particular use case we have (as the model constructor requires a frame, which can't be resurrected from an instance's state_dict alone).  I've removed the default value of copy_fn from DataParallelizedModule.__init__.